### ISSUE TITLE
Have Legends Dynamically Change to only Show Visible Options

### ISSUE THREAD

**Issue by emmahodcroft** [nextstrain/auspice#1109]
**Context**  
Currently for some builds, like nextstrain.org/ncov, the legends are currently so long that they are not very useful:
![image](https://user-images.githubusercontent.com/14290674/81193198-5ff88400-8fbb-11ea-8936-9fb8fb8c76f9.png)
(Countries shown, but valid for many of the selections)

@MoiraZuber wrote a tweet for a recent push to nextstrain.org where she had only 'new' sequences selected via filters. She manually adjusted the legend so that it only showed entries for the nodes that were currently visible, with the filters selected. 
<img width="739" alt="pic_1 (1)" src="https://user-images.githubusercontent.com/14290674/81193378-99c98a80-8fbb-11ea-9ccf-e1316e7a3e67.png">
('division' coloring on the nextstrain.org/ncov/north-america build shown)

This makes the legend suddenly incredibly useful again!

**Description**   
When filters are applied, only values that are present on the currently visible nodes should be shown on the legend. Note that the colours do not *change* - just the number of things shown in the legend is reduced to only be the tips visible.

However, for traits where ancestral reconstruction has taken place (like 'countries' on the main nextstrain.org/ncov build), and branches are coloured, it should be weighed up whether the traits on visible *branches* should also be included in the traits shown on the legend.  Happy to discuss the pros and cons of this with anyone interested in implementing - my gut feeling is that it would be worth including branches that lead to visible nodes (these are already identified in the visualisation by being thicker), but if difficult to implement, or on average introduces too many other traits to the legend, this might not be worth it.

**Comment by mingir2** [comment]
@emmahodcroft @jameshadfield 

<img width="1418" alt="Screen Shot 2020-05-10 at 9 02 06 PM" src="https://user-images.githubusercontent.com/40753283/81517746-842ccb80-9301-11ea-91f6-8949ce22400b.png">
<img width="1412" alt="Screen Shot 2020-05-10 at 9 02 31 PM" src="https://user-images.githubusercontent.com/40753283/81517767-93137e00-9301-11ea-9997-1f61a39c8e17.png">

So, I've looked into this a bit and I was able to get it working within the UI component. But I need some help with understanding the application architecture to integrate it directly into the colorScale.legendValues pipeline.

Basically, it works by filtering all visible nodes with tree.nodes and tree.visibility. With the visible nodes, we can identify visible legend values with colorScale.legendValues and a set of node_attrs.value.
```js
  /*
   * filter tree nodes by visibility to create a new set of visible legend values.
   * returns new legend values, which was filtered from the actual legend values with the new set.
   * TODO: 
   *  1. Replace validColorBy with a global reference if there is one.
   *  2. Move this function to where it can work better.
   */
  getLegendValues = () => {
    const validColorBy = ["clade_membership", "location", "country", "division", "region", "host"];
    let legendValues = this.props.colorScale.legendValues;

    if (validColorBy.includes(this.props.colorBy)) {
      const visibleNodes = this.props.tree.nodes.filter((node, i) => (
        this.props.tree.visibility[i] && node.node_attrs[this.props.colorBy]
      ));
      const filterValues = new Set(visibleNodes.map(node => node.node_attrs[this.props.colorBy].value));
      legendValues = legendValues.filter(value => value !== undefined && filterValues.has(value));
    }
    
    return legendValues;
  }
```
[source](https://github.com/mingir2/auspice/commit/05e5ac5e7a0ab32bad55b7e94c07097be00daf86)

However, this shouldn't be done in the UI component (legend.js). Instead, it should be done via action/dispatch, but I'm not sure where it belongs and at what stage. I tried to integrated it into the calcColorScale() function, but it wasn't being called when a new filter was applied.

So, I'm not entirely sure if I should add another dispatch() line to the applyFilter() in actions/tree.js, or there's a better way to do this that I'm not aware of.

**Comment by jameshadfield** [comment]
Thanks @mingir2 -- this looks good. In terms of state, we've currently got `reduxState.controls.colorScale.legendValues` so the most straightforward would be to create `reduxState.controls.colorScale.visibileLegendValues` and then make any actions which change tree visibility to also recalculate this. I've added a commit which does this at https://github.com/nextstrain/auspice/commit/7d0f9480398fb94b73af9f10edd46860119a6d16 which I hope will help you. There are some edge conditions which need looking at, as well as handling different scales appropriately. Happy to look at a PR no matter how incomplete!

**Comment by mingir2** [comment]
@jameshadfield Thank you so much! I'll work on it week and hopefully get back to you soon with a PR.

**Comment by jameshadfield** [comment]
Having thought about this a bit more, I think this functionality is desperately needed for categorical and ordinal scales, but that we do not want it for boolean and continuous ones. The implementation I prototyped in 7d0f948 doesn't need overhauling, rather `createVisibleLegendValues` can simply return a `slice` of the legend items for boolean & continuous scales (this is trivial in the scheme of the computation we're doing). 

The first of the following two screenshots is clearly problematic, but I do not think the second one should have a more restricted legend.

![image](https://user-images.githubusercontent.com/8350992/83221438-ee7eb080-a1c9-11ea-8bd6-24a5cd626bf3.png)

![image](https://user-images.githubusercontent.com/8350992/83221458-fd656300-a1c9-11ea-9a76-247daa9ed6ff.png)

**Comment by emmahodcroft** [comment]
Good thinking James. I'd second this.

**Comment by mingir2** [comment]
@jameshadfield I agree that boolean and continuous scales shouldn't change based on visibility. 

By the way, I'm not sure what's going on with the first image with the duplicate color. What's the steps to reproduce it?

**Comment by mingir2** [comment]
@jameshadfield I addressed the race condition issue and modified the ```createVisibleLegendValues``` function based on our discussion on this thread. Please take a look at the PR #1150.

**Comment by jameshadfield** [comment]
> By the way, I'm not sure what's going on with the first image with the duplicate color. What's the steps to reproduce it?

Auspice will do this automatically if there are more values than there are colours available. See https://nextstrain.org/zika?c=author for an example of this
